edb/server/compiler/compiler.py

```diff
@@ -2194,8 +2228,20 @@ def compile_in_tx(
         protocol_version: Tuple[int, int],
         inline_objectids: bool = True,
         json_parameters: bool = False,
+        expect_rollback: bool = False,
     ) -> Tuple[dbstate.QueryUnitGroup, dbstate.CompilerConnectionState]:
-        state.sync_tx(txid)
+        if (
+            expect_rollback and
+            state.current_tx().id != txid and
+            not state.can_sync_to_savepoint(txid)
+        ):
+            # This is a special case when COMMIT MIGRATION fails, the compiler
+            # doesn't have the right transaction state, so we just roll back.
+            return (
+                self.try_compile_rollback(source, protocol_version)[0], state
+            )
```

Comment:
This is not the only solution - alternatively, we could rollback the transaction when `COMMIT MIGRATION` was failing.